source/rust_verify/src/external.rs

```diff
@@ -225,6 +239,49 @@ fn opts_in_to_verus(eattrs: &ExternalAttrs) -> bool {
 }
 
 impl<'a, 'tcx> VisitMod<'a, 'tcx> {
+    fn visit_opaque(&mut self, opaque: &'tcx OpaqueTy<'tcx>) {
+        rustc_hir::intravisit::walk_opaque_ty(self, opaque);
+        let parent_def_id = {
+            match opaque.origin {
+                rustc_hir::OpaqueTyOrigin::FnReturn { parent, .. } => parent,
+                rustc_hir::OpaqueTyOrigin::AsyncFn { parent, .. } => parent,
+                rustc_hir::OpaqueTyOrigin::TyAlias { parent, .. } => parent,
+            }
+        };
+
+        let attrs = self.ctxt.tcx.get_attrs_unchecked(parent_def_id.into());
+        let eattrs = match self.ctxt.get_external_attrs(attrs) {
+            Ok(eattrs) => eattrs,
+            Err(err) => {
+                self.errors.push(err);
+                return;
+            }
+        };
+
+        let state_for_this_item =
+            { if eattrs.external { VerifState::External } else { VerifState::Verify } };
+
+        let verif = if state_for_this_item == VerifState::Verify {
+            VerifOrExternal::VerusAware {
+                module_path: self.module_path.clone(),
+                const_directive: eattrs.size_of_global || eattrs.item_broadcast_use,
+                external_body: eattrs.external_body,
+            }
+        } else {
+            let path_opt = def_id_to_vir_path_option(
+                self.ctxt.tcx,
+                Some(&self.ctxt.verus_items),
+                opaque.def_id.into(),
+            );
+            let path_string = match &path_opt {
+                Some(path) => vir::ast_util::path_as_friendly_rust_name(&path),
+                None => format!("{:?}", opaque.def_id,),
+            };
+            VerifOrExternal::External { path: path_opt, path_string, explicit: false }
+        };
+
+        self.opaque_tys.push(OpaqueDef { id: opaque.def_id, verif: verif });
+    }
```

Comment:
The significance of these additions to `external.rs` is eluding me. Why does `CrateItems` need an `opaque_tys` field? Why is the existing traversal logic not sufficient for handling opaque types?

The computation for an opaque type's `state_for_this_item` only depends on the `eattrs`, but not on the `state` of the parent node, which is different than the traversal for other items, which propagates the state inwards. Is there a reason for this, and if so, what is it?

I would think you'd want the externalness of an opaque type to be the same as the externalness of the corresponding function, but due to the difference, but I don't think that it's going to be the case here, considering the various differences between this and the `visit_general` function.